src/Db.php

```diff
@@ -17,8 +17,10 @@ public static function getInstance(): Db {
   private function __construct() {
     $this->config = parse_ini_file($this->settings_file);
     $options = array(
-      'idle_timeout_micros' => 200000,
+      'idle_timeout_micros' => 2000000,
       'expiration_policy' => 'IdleTime',
+      'per_key_connection_limit' => 20,
```

Comment:
> `per_key_connection_limit' => 20`

**Can you explain the justification for lowering this from the default value?**

The default is `50` connections.  Lowering this to `20` means that if any additional connections are requested after 20 exist in the pool, an exception will be thrown and the connection will not be created.  _It should be noted that this value is specific to number of allowed connections per MySQL server hostname, port, database, and username._

In optimal conditions, lowering this value to `20` would not cause any issue.  In fact, when things are running smoothly, the maximum number of connections in the pool is only `2`.  Currently a full AJAX refresh of the gameboard requests just over `50` connections, but the project maintains the default of `setReusable` so connections in the pool that are idle will be reused (or safely recreated if they have been destroyed).

The potential issue arises when those connections in the pool are not idle.  Imagine a scenario where there is a high load on the database and the gameboard AJAX refresh is requesting _just over_ `50` connections, and none of the connections in the pool are available for reuse yet (they are still requesting/retrieving data).

To be honest, given the way the project is structured (especially, with the caching) a connection pool with `50` or even `20` connections seems highly unlikely, and would likely mean other timeouts are being hit or things are failing.  So I do not think there is much risk in lowering this value.  However, there doesn't seem to be much benefit, especially with the ease of increasing the MySQL connection limit (`max_connections`).  Again, I believe erring on the side of slow data over no data is appropriate.

Do you have any test data that provides credence to lowering this particular value?

Reply:
I think 50 connections is too much. I agree that requests are going to fail, but I saw that 20 was a number were requests didn't fail and I didn't risk taking all of my available connections (which were around 1500).

I think this is a "hack" and we should improve the reuse of connections, because I was some requests creating up to 17 connections for a SINGLE requests. When you have more than 200 concurrent users playing, that's a LOT. I preferred returning some errors than saturating my MySQL to all my users.